### ISSUE TITLE
LeoFS Gateway - graceful configuration reload 

### ISSUE THREAD

**Issue by pannon** [leo-project/leofs#673]
Is there a way to gracefully reload gateway configuration without restarting the process itself?

We are modifying the http_custom_header.conf file quite often, would like to avoid gateway restarts if possible to reload the new headers.

**Comment by yosukehara** [comment]
@pannon 

I've checked the current implementation, which only once load `http_custom_header.conf` at launching a LeoGateway node:
- [leo_gateway_http_commons:start/1](https://github.com/leo-project/leofs/blob/master/apps/leo_gateway/src/leo_gateway_http_commons.erl#L77)

How often does you reload the configuration? After receiving your answer, we will consider whether we need to implement reloading the configuration feature or not.

**Comment by pannon** [comment]
At this stage we are reloading every couple of days. In the testing environments multiple times a day.

**Comment by mocchira** [comment]
@pannon 
Thank you for replying.
While still not decided to do what you've requested yet, we are considering your request positively :)

@yosukehara 
Just in case,
If we go forward this issue, there are a few ways to realize.

- Add a reload command to leo_gateway script
It can be realized through nodetool
```shell
### reload .conf files without restarting
$ ./package/leo_gateway/bin/leo_gateway reload
```

- Handle a specific OS signal (typically SIGHUP/SIGUSR1)
It will be able to realize through Signal Service: https://github.com/erlang/otp/pull/1315 
```shell
### reload .conf files without restarting
$ kill -HUP ${leo_gateway_pid} 
```

**Comment by pannon** [comment]
Thanks! Maybe something along the lines of `nginx -t && nginx -s reload`.
Would be nice to have a way for quick config syntax test before reloading, especially when things are plugged into automation frameworks like consul-template, Puppet, Salt, Chef, etc.

This would be a safe choice for most use cases and users.

**Comment by mocchira** [comment]
WIP.

Note: Since signal handling in Erlang has been supported from OTP20, we go for the nodetool way.

**Comment by mocchira** [comment]
@pannon It's been a long time since you filed this issue. We've finally implemented this feature and it will be shipped with upcoming 1.4.1 (probably today) so please give it a try once it gets released.
```shell
# reload
$ ./package/leo_gateway/bin/leo_gateway reload_http_conf
# conf test
$ ./package/leo_gateway/bin/leo_gateway test_http_conf
```
Feel free to reopen if you find any problems.
